src/ui/widgets/XYPlot/xyPlotOptions.ts

```diff
@@ -0,0 +1,259 @@
+        ticks: "outside"
+      },
+      tickformat: axis.scaleFormat
+    };
+    // Only add title labels if they exist
+    if (axis.axisTitle !== "") {
+      _newAxis.title = {
+        text: axis.axisTitle,
+        standoff: 0
+      };
+      _newAxis.titlefont = {
+        family: font ? font.fontFamily : "Liberation sans, sans-serif",
+        size: font.fontSize ? font.fontSize : 12
+      };
+    }
+    _newAxis.range = [axis.minimum, axis.maximum];
+    _newAxis.autorange = false;
+
+    if (axis.logScale) _newAxis.type = "log";
+
+    // More than 2 axes, determine position and location of axes
+    if (idx > 1) {
+      _newAxis.side = axis.leftBottomSide ? "left" : "right";
+      _newAxis.overlaying = axis.yAxis ? "y" : "x";
+      _newAxis.position = 0;
+      _newAxis.anchor = "free";
+      // Only shift to add space for 2nd y-axis if it is overlapping and visible
+      if (axis.visible)
+        formattedAxes[0].domain = [0.08 * (axesList.length - 1), 1];
+    }
+    formattedAxes.push(_newAxis);
+    idx += 1;
+  }
+  return formattedAxes;
+}
+
+/**
+ * Calculates the axis range within the maximum
+ * and minimum limits using the highest and lowest
+ * data values. Defaults to range specified from file
+ * if calculated limits are larger.
+ * @param oldAxis options object read from file
+ * @param newAxis plotly axis options object
+ * @param dataSet array of PV data
+ * @param index the index of the axis
+ * @returns modified axis options object
+ */
+export function calculateAxisLimits(
+  oldAxis: Axis,
+  newAxis: NewAxisSettings,
+  dataSet: any[]
+): NewAxisSettings {
+  // If autoscale not enabled, do not rescale axes
+  if (!oldAxis.autoScale) return newAxis;
+  // Find axis limits
+  const [axMin, axMax]: [number, number] = newAxis.range;
+  // Determine if we are using x or y data
+  let dataType = "x";
+  if (oldAxis.index > 0 && oldAxis.yAxis) dataType = "y";
+  // Combine all data we need into an array of arrays
+  const allData = dataSet.map(val => val[dataType]);
+  // Find max and minimum value from all the traces and round to 1sf
+  const actualMin = Math.min(...[].concat(...allData));
+  const actualMax = Math.max(...[].concat(...allData));
+  let min = roundValue(Math.min(actualMin), 0);
+  let max = roundValue(Math.max(actualMax), 1);
+  // If number is within autoscale threshold, just use axis limits
+  const threshold = 1 - (max - min) / (axMax - axMin);
+  if (
+    oldAxis.autoScaleThreshold &&
+    -oldAxis.autoScaleThreshold < threshold &&
+    threshold < oldAxis.autoScaleThreshold
+  ) {
+    min = axMin;
+    max = axMax;
+  }
+  newAxis.range = [min, max];
+  return newAxis;
+}
+
+/**
+ * Round a decimal value up or down to the nearest
+ * significant figure.
+ * @param value decimal value to round
+ * @param mag magnitude to round to
+ * @param roundType 0 to round down, 1 to round up
+ * @returns rounded number
+ */
+export function roundValue(value: number, roundType: number): number {
```

Comment:
Pretty surprised you had to implement this yourself.

Reply:
Still, probably better to do this than add yet another dependency.

Reply:
The TS `.floor()` and `.ceil()` methods always round to integers. While there are methods that can round decimals to a specified significant figure after the dot (e.g. `toFixed()`), there isn't a way to specify whether it should round up or down. I was also surprised that something like this doesn't seem to exist already in TS...

Reply:
A useful function - perhaps it could be moved to a common location? Maybe we should create a new 'utils' file with some of these methods in?